### PR TITLE
[3/?] - lnwire: add new TLV record types + messages for taproot chans

### PR DESCRIPTION
**Pull request by Roasbeef**
## Change Description

This PR was split off of #6877. 

This PR adds the set of new messages and TLV types for the current spec draft of simple taproot channels. None of the messages in this PR are used yet, as the goal is to split up the current mega PR into a series of digestible PRs. 

The main change to the _existing_ fields is the overloading of the `lnwire.Sig` type. This message is now updated to support encoding/decoding both ECDSA and schnorr signatures. We do this via the addition of a new type enum, which is consulted when converting between an `lnwire.Sig` and an `input.Signature`. 

The other changes are adding the new partial sig, and nonce types to the relevant set of messages. Local/verification nonces are exchange up front by both sides. When a signature needs to be generated, a signing/remote nonce is generated prior, which is then included along with the nonce. 
### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [ ] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [ ] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: lnwire/partial_sig.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+// signature. Signers can use this trait to mix in additional contextual data
+// such as the commitment txn itself into their nonce generation function.
+//
+// The final signature is 98 bytes: 32 bytes for the S value, and 66 bytes for
+// the public nonce (two compressed points).
+type PartialSigWithNonce struct {
+	PartialSig
+
+	// Nonce is the 66-byte musig2 nonce.
+	Nonce Musig2Nonce
```
Comment (ellemouton) [inline lnwire/partial_sig.go:116]: The `Musig2Nonce` type is only introduced in the next commit. So just gotta swap these two commits for this to compile

File: lnwire/partial_sig.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+)
+
+const (
+	// PartialSigLen is the length of a musig2 partial signature.
+	PartialSigLen = 32
+
+	// PartialSigRecordType is the type of the tlv record for a musig2
+	// partial signature. This is an _even_ type, which means it's required
+	// if included.
+	PartialSigRecordType = 6
```
Comment (ellemouton) [inline lnwire/partial_sig.go:18]: Feels like all these types should be declared in 1 spot so it is easy to see which types are taken etc? cause wont the type number maybe vary depending on which message it is in?

Also, should we be strict and force the type here to do `=  tlv.Type(4)` rather? (same for `NonceRecordType`)

File: lnwire/open_channel.go [hunk 3]
```diff
@@ -141,6 +141,13 @@ type OpenChannel struct {
 	// type.
 	LeaseExpiry *LeaseExpiry
 
+	// LocalNonce is an optional field that transmits the
+	// local/verification nonce for a party. This nonce will be used to
+	// verify the very first commitment transaction signature.  This will
+	// only be populated if the simple taproot channels type was
+	// negotiated.
+	LocalNonce *Musig2Nonce
```
Comment (ellemouton) [inline lnwire/open_channel.go:149]: I think this is out of sync with the bolt proposal? There it is specified that we add both a verification nonce _and_ a signing nonce to this message and others

File: lnwire/funding_created.go [hunk 4]
```diff
@@ -26,6 +27,13 @@ type FundingCreated struct {
 	// transaction.
 	CommitSig Sig
 
+	// PartialSig is used to transmit a musig2 extended partial signature
+	// that also carries along the public nonce of the signer.
+	//
+	// NOTE: This field is only populated if a musig2 taproot channel is
+	// being signed for. In this case, the above Sig type MUST be blank.
+	PartialSig *PartialSigWithNonce
```
Comment (ellemouton) [inline lnwire/funding_created.go:35]: ok, I see now that the Signing nonce that was originally sent in open-channel (as described in the bolt proposal) is now sent here? 

just out of interest, what was the reason for this change? cause from my understanding, If we sent sent this nonce in open-channel then here we could reuse the `signature` field to send the partial sig (with some padding)

Comment (ellemouton) [reply lnwire/funding_created.go:35]: ah - Is the reason that there are fewer bytes to send across on connection reestablish?

File: lnwire/extra_bytes.go [hunk 5]
```diff
@@ -90,6 +90,10 @@ func (e *ExtraOpaqueData) ExtractRecords(recordProducers ...tlv.RecordProducer)
 		records = append(records, producer.Record())
 	}
 
+	// Ensure that the set of records are sorted before we attempt to
+	// decode from the stream, to ensure they're canonical.
+	tlv.SortRecords(records)
+
```
Comment (ellemouton) [inline lnwire/extra_bytes.go:96]: nice! 
btw, any reason this is not done automatically within the `tlv` package before `Encode` or on `tlv.NewStream`?

**Review by ellemouton — COMMENTED** [review]
First pass done! Looking slick 🔥 

My main initial confusion was due to this being a bit out of sync with the current spec proposal doc. 

But other than that, looks good & was easy to follow

File: lnwallet/chancloser/chancloser.go [hunk 6]
```diff
@@ -785,16 +785,18 @@ func (c *ChanCloser) proposeCloseSigned(fee btcutil.Amount) (*lnwire.ClosingSign
...
-	parsedSig, err := lnwire.NewSigFromSignature(rawSig)
-	if err != nil {
-		return nil, err
+	// We'll note our last signature and proposed fee so when the remote
+	// party responds we'll be able to decide if we've agreed on fees or
+	// not.
+		parsedSig, err = lnwire.NewSigFromSignature(rawSig)
+		if err != nil {
+			return nil, err
+		}
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwallet/chancloser/chancloser.go:795]: Rebased issue, fixed.

File: lnwire/signature.go [hunk 7]
```diff
@@ -23,14 +18,74 @@ var (
...
+	bytes [64]byte
+
+	sigType sigType
+}
+
+// ForceSchnorr forces the signature to be interpreted as a schnorr signature.
+// This is useful when reading an HTLC sig off the wire for a taproot channel.
+// In this case, in order to obtain an input.Signature, we need to know that
+// the sig is a schnorr sig.
+func (s *Sig) ForceSchnorr() {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/signature.go:49]: So this'll be used later in the series. The main place we need to use this is for the second-level HTLC sigs (we reuse the `[64]byte` sig field). When we read things off the wire, unless we want to make the parsing contextual (we'd check to see if other TLV fields are present to _then_ coerce it), we need to do this as an application level check.

File: lnwire/partial_sig.go (hunk 2 above)
Comment (Roasbeef) [reply lnwire/partial_sig.go:18]: > Feels like all these types should be declared in 1 spot so it is easy to see which types are taken etc? cause wont the type number maybe vary depending on which message it is in?

Yeah good question, technically the TLV namespace is sort of scoped to each individual message. Eg: all the mesages can use a field of 1 and it won't collide since the message itself serves as a local namespace. In the current spec draft, we end up creating a psuedo-global namespace that applies to all these existing messages. 

I modified the declaration to declare the type `tlv.Type = 4`.

**Comment by Roasbeef** [comment]
@ellemouton I merged the PR to my BOLT fork that this was baed on. Things should be synced now.

File: lnwire/open_channel.go (hunk 3 above)
Comment (Roasbeef) [reply lnwire/open_channel.go:149]: Commented below, this integrates the nonce modifications proposed by the LDK team.

File: lnwire/funding_created.go (hunk 4 above)
Comment (Roasbeef) [reply lnwire/funding_created.go:35]: The general idea is that we can defer sending the nonce we use to sign to the very last step. This lets the signer mix in more randomness, and also means they don't need to commit to session creation until there's a state to create/accept. 

It also makes chan reest simpler as well, since you're already generating the nonce at signing time, so you can either retransmit what you had, or just generate a new nonces anew.

File: lnwire/extra_bytes.go (hunk 5 above)
Comment (Roasbeef) [reply lnwire/extra_bytes.go:96]: Good q...I think the intention was that the package detected an invalid ordering rather than catch it at that instance. We can def revisit that assumption tho. IIRC, we added these sorting routines once we started to display/accept the TLV fields on the RPC interface.

**Comment by Roasbeef** [comment]
Pushed up a new version. Things should be building now, and you can use my spec PR as a base for the review. 

PTAL @ellemouton @yyforyongyu.

**Comment by Roasbeef** [comment]
Fixed some more API test/compilation issues in the latest diff.

File: lnwire/fuzz_test.go [hunk 8]
```diff
@@ -387,6 +387,7 @@ func FuzzNodeAnnouncement(f *testing.F) {
 			first.Signature.RawBytes(),
 			second.Signature.RawBytes(),
 		) {
+
```
Comment (ellemouton) [inline lnwire/fuzz_test.go:390]: nit: much nothing

**Review by ellemouton — APPROVED** [review]
LGTM 🚰 🥕

**Comment by ellemouton** [comment]
Ah - actually - all itest are failing..

**Comment by Roasbeef** [comment]
@ellemouton fixed! Was using the wrong sig parsing for the `RPCKeyRing.SignMessage` method.

**Comment by Roasbeef** [comment]
Review ping @yyforyongyu.

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Smooth! And sorry that missed the ping🤦🏻 My main question is the addition of the shutdown nonce, not sure about I understand it. Other than that a few typos and nits, then this is good to go!

**Comment by Roasbeef** [comment]
Note that I've changed the base branch to the new taproot staging branch as mentioned in: https://github.com/lightningnetwork/lnd/issues/6691#issuecomment-1452827837

File: lnwire/features.go [hunk 9]
```diff
@@ -221,6 +221,14 @@ const (
 	// TODO: Decide on actual feature bit value.
 	ScriptEnforcedLeaseOptional FeatureBit = 2023
 
+	// SimpleTaprootChannelsRequred is an required bit that indicates the
+	// node is able to create unadvertised taproot-native channels.
+	SimpleTaprootChannelsRequired = 80
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/features.go:226]: Nope, good call out, removed.

File: lnwire/signature.go [hunk 10]
```diff
@@ -23,14 +18,74 @@ var (
...
+// Today this can either be an ECDSA sig or a schnorr sig. Both of these can
+// fit cleanly into 64 bytes.
+type sigType uint
+
+const (
+	// sigTypeECDSA represents an ECDSA signature.
+	sigTypeECDSA sigType = iota
+
+	// sigTypeSchnorr represents a schnorr signature.
+	sigTypeSchnorr
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/signature.go:31]: If we want the entire commit to compile, then we need all the other API changes coupled in as well.

File: lnwire/signature.go [hunk 11]
```diff
@@ -23,14 +18,74 @@ var (
...
+
+// Sig is a fixed-sized ECDSA signature or 64-byte schnorr signature. For the
+// ECDSA sig, unlike Bitcoin, we use fixed sized signatures on the wire,
+// instead of DER encoded signatures. This type provides several methods to
+// convert to/from a regular Bitcoin DER encoded signature (raw bytes and
+// *ecdsa.Signature).
+type Sig struct {
+	bytes [64]byte
+
+	sigType sigType
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/signature.go:42]: If we did that, then for all the instances where either sig type can be used (closing signed, commit_sig, etc) we'd need to add this other type along side things, then have other methods to check which was set, then extract the sig from that. 

There's another option where we make those messages generic (type params), but then this bleeds into all the APIs a bit, as any area we're storing one of those fields needs to mind the type param (`ChanCloser[schnorr.signature]`). 

This approach takes into account that they're actually the same set of bytes and are serialized identically on the wire (just the raw 64 bytes). The main quirk is the `ForceSchnorr` method. We need this as in certain situations, we need to be able to extract a schnorr signature from an opaque `lnwire.Sig` type. This comes up when we're going to validate a second level HTLC, and we know it's a taproot channel, so we need a schnorr sig. 

Alternatively, we can add a new `lnwire.Sig` interface, then have an implementation for ECDSA and schnorr. I started doing this initially, but realized that later up the stack (when you need to actually use the sigs), we still need methods to obtain the concrete types (we go indirectly thru `input.Signature` to do that atm).  

So to summarize, this approach is closer to how things are in the spec (we overload the `[64]byte` on the wire, and context is needed to obtain the proper concrete type). We can try generics (type params), but that complicates the API for seemingly minimal gain. We can also try to make a new interface for `lnwire.Sig` here similar to `input.Signature` -- my gut tells me YAGNI tho, as this is only the second new signature type in over a decade of Bitcoin's existence.

File: lnwire/musig2.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+func (m *Musig2Nonce) Record() tlv.Record {
+	return tlv.MakeStaticRecord(
+		NonceRecordType, m, musig2.PubNonceSize, nonceTypeEncoder,
+		nonceTypeDecoder,
+	)
+}
+
+// nonceTypeEncoder is a custom TLV encoder for the Musig2Nonce type.
+func nonceTypeEncoder(w io.Writer, val interface{}, buf *[8]byte) error {
+	if v, ok := val.(*Musig2Nonce); ok {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/musig2.go:30]: Seems like minor preference? Everywhere else I've used TLV, I used this pattern. See the channel type (one of the first instances we used it): https://github.com/lightningnetwork/lnd/blob/master/lnwire/channel_type.go#L34-L44

File: lnwire/partial_sig.go [hunk 13]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+// partialSigWithNonceTypeEncoder encodes 98-byte musig2 extended partial
+// signature as: s {32} || nonce {66}.
+func partialSigWithNonceTypeEncoder(w io.Writer, val interface{},
+	buf *[8]byte) error {
+
+	if v, ok := val.(*PartialSigWithNonce); ok {
+		sigBytes := v.Sig.Bytes()
+		if _, err := w.Write(sigBytes[:]); err != nil {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/partial_sig.go:147]: Not sure if we're looking at the same diff, but `partialSig.Encode` ends up calling this.

File: lnwire/funding_created.go [hunk 14]
```diff
@@ -63,10 +80,36 @@ func (f *FundingCreated) Encode(w *bytes.Buffer, pver uint32) error {
...
+		return err
+	}
+
+	// Set the corresponding TLV types if they were included in the stream.
+	if val, ok := typeMap[PartialSigWithNonceRecordType]; ok && val == nil {
+		f.PartialSig = &partialSig
+	}
+
+	if len(tlvRecords) != 0 {
+		f.ExtraData = tlvRecords
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/funding_created.go:109]: So this is a pattern left over from when we needed to start storing the opaque TLV for gossip messages: if we needed to encode/decode TLV, then we also set the `ExtraData` field as it might contain TLV keys we don't actually know about.

File: lnwire/funding_locked.go [hunk 15]
```diff
@@ -76,6 +90,13 @@ func (c *FundingLocked) Decode(r io.Reader, pver uint32) error {
 	if val, ok := typeMap[AliasScidRecordType]; ok && val == nil {
 		c.AliasScid = &aliasScid
 	}
+	if val, ok := typeMap[NonceRecordType]; ok && val == nil {
+		c.LocalNonce = &localNonce
+	}
+
+	if len(tlvRecords) != 0 {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/funding_locked.go:97]: See comment above, this ensures that if we don't understand certain types then we still end up writing them to disk for future broadcast. 

IIUC, we also need to do this as otherwise the quick check tests start failing (raw bytes don't match the struct and other way around). 

Just to sync up on mental models here: this sets the `ExtraData` to what it would look like right after the call to `Encode`. Otherwise, the struct that had `Encode` called on it would have `ExtraData` populated to raw TLV, while the struct that had `Decode` on it wouldn't. This means a strict equality check fails.

File: lnwire/revoke_and_ack.go [hunk 16]
```diff
@@ -54,19 +60,52 @@ var _ Message = (*RevokeAndAck)(nil)
...
 		c.Revocation[:],
 		&c.NextRevocationKey,
-		&c.ExtraData,
 	)
+	if err != nil {
+		return err
+	}
+
+	var tlvRecords ExtraOpaqueData
+	if err := ReadElements(r, &tlvRecords); err != nil {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/revoke_and_ack.go:73]: Good question...🤔

Mostly just following the pattern of other areas we use TLV like `OpenChannel` (for the channel type field): https://github.com/lightningnetwork/lnd/blob/master/lnwire/open_channel.go#L282-L288

File: lnwire/partial_sig.go [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+		partialSigTypeEncoder, partialSigTypeDecoder,
+	)
+}
+
+// partialSigTypeEncoder encodes a 32-byte musig2 partial signature as a TLV
+// value.
+func partialSigTypeEncoder(w io.Writer, val interface{}, buf *[8]byte) error {
+	if v, ok := val.(*PartialSig); ok {
+		sigBytes := v.Sig.Bytes()
+		if _, err := w.Write(sigBytes[:]); err != nil {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/partial_sig.go:49]: Good idea, done!

File: lnwire/partial_sig.go [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+		var sBytes [32]byte
+		if _, err := io.ReadFull(r, sBytes[:]); err != nil {
+			return err
+		}
+
+		var s btcec.ModNScalar
+		s.SetBytes(&sBytes)
+
+		var nonce [66]byte
+		if _, err := io.ReadFull(r, nonce[:]); err != nil {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/partial_sig.go:182]: Used `DBytes32` above, but this serialization wants just 66 bytes, so can't use `DVarBytes`.

File: lnwire/funding_locked.go [hunk 19]
```diff
@@ -95,12 +116,16 @@ func (c *FundingLocked) Encode(w *bytes.Buffer, pver uint32) error {
 	}
 
 	// We'll only encode the AliasScid in a TLV segment if it exists.
+	var recordProducers []tlv.RecordProducer
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/funding_locked.go:119]: Done.

File: lnwire/commit_sig.go [hunk 20]
```diff
@@ -36,6 +38,13 @@ type CommitSig struct {
 	// transaction should be signed.
 	HtlcSigs []Sig
 
+	// PartialSig is used to transmit a musig2 extended partial signature
+	// that also carries along the public nonce of the signer.
+	//
+	// NOTE: This field is only populated if a musig2 taproot channel is
+	// being signed for. In this case, the above Sig type MUST be blank.
+	PartialSig *PartialSigWithNonce
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/commit_sig.go:46]: In the future changes, this value is always converted to a concrete type we use. There's no big diff between them, other than that for co-op close we only have a single message to sign.

File: lnwire/shutdown.go [hunk 21]
```diff
@@ -3,8 +3,56 @@ package lnwire
...
 	"bytes"
 	"io"
+
+	"github.com/btcsuite/btcd/btcec/v2/schnorr/musig2"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+const (
+	// ShutdownNonceRecordType is the type of the shutdown nonce TLV record.
+	ShutdownNonceRecordType = 8
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/shutdown.go:13]: The LDK reviewers wanted to use distinct types here (compared to that sent in the chan open) messages citing a diff context/semantics.

File: lnwire/lnwire_test.go [hunk 22]
```diff
@@ -544,6 +586,16 @@ func TestLightningWireProtocol(t *testing.T) {
 				return
 			}
 
+			// 1/2 chance to attach a partial sig.
+			if r.Intn(2) == 0 {
```
[the remark this answers is not in the document]
Comment (Roasbeef) [reply lnwire/lnwire_test.go:590]: If we want to go in this direction, then I think we'd overhaul all the tests to do the same. I originally added both everywhere, as it showed some bugs that would come up when only some of the TLV fields were present.

**Comment by Roasbeef** [comment]
New version pushed, PTAL.

**Comment by Roasbeef** [comment]
Fixed a linter issue. Forces us to have a new line in the fuzz pacakge.

File: lnwire/fuzz_test.go
Comment (yyforyongyu) [inline lnwire/fuzz_test.go:1]: this commit needs to be squashed

**Review by yyforyongyu — COMMENTED** [review]
The last commit needs to be squashed, otherwise this is good to go⚡️
